### PR TITLE
Dirty Reads

### PR DESCRIPTION
**Pull request by lamai93**
Fixes arangodb/team-clifton-planning#47.

[Docs](https://github.com/arangodb/documents/blob/master/Driver/DirtyReads.md)

### CODE REVIEW

**Comment by lamai93** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/1268/

File: context.go [hunk 1]
```diff
@@ -137,6 +138,14 @@ func WithWaitForSync(parent context.Context, value ...bool) context.Context {
 	return context.WithValue(contextOrBackground(parent), keyWaitForSync, v)
 }
 
+// WithAllowDirtyReads is used in an active failover scenario to allow reads from the follower.
```
Comment (ewoutp) [inline context.go]: scenario -> deployment

File: context.go [hunk 2]
```diff
@@ -240,6 +251,19 @@ type contextSettings struct {
 	JobIDResponse            *string
 }
 
+// loadContextResponseValue loads generic values from the response and puts it into variables specified
+// via context values.
+func loadContextResponseValues(cs contextSettings, resp Response) {
+	// Parse potential dirty read
+	if cs.AllowDirtyReads && cs.DirtyReadFlag != nil {
```
Comment (ewoutp) [inline context.go]: Don't think you need to consider `AllowDirtyReads` here. `DirtyReadFlag != nil` is sufficient.

File: context.go [hunk 3]
```diff
@@ -240,6 +251,19 @@ type contextSettings struct {
...
+		if dirtyRead := resp.Header("X-Arango-Potential-Dirty-Read"); dirtyRead != "" {
+			*cs.DirtyReadFlag = (dirtyRead == "true")
```
Comment (ewoutp) [inline context.go]: Are you sure it is alway `true`? If not consider using `strconv.ParseBool`

File: cursor_impl.go [hunk 4]
```diff
@@ -139,19 +141,31 @@ func (c *cursor) ReadDocument(ctx context.Context, result interface{}) (Document
 	// Force use of initial endpoint
 	ctx = WithEndpoint(ctx, c.endpoint)
 
+	// Concerns: the `wasDirtyRead` flag is updated only when a new batch is fetched.
+	// From a user perspective this results in strange and inconsistent behaviour. Fetching a batch is an implementation detail.
+	// Maybe we should not support updating this flag here or set it actively in every call, regardless of fetching a new batch.
+	//
+	// if dirty reads have been set before and the current context does not
+	// allow dirty reads, we have to allow it here.
+	if c.allowDirtyReads && ctx.Value(keyAllowDirtyReads) == nil {
```
Comment (ewoutp) [inline cursor_impl.go]: Probably `&&` should be `||` here

**Pull request merged**